First draft reviews (#9)
bcgov/platform-developer-docs: platform-storage.md

page purpose:

audience:
---

# Platform storage

We use several different storage technologies on the OpenShift platform. Currently, we have three types of persistent storage available:

* **FILE**: File storage is an all-purpose storage type that can be attached to one or more containers. It's the recommended storage class for most general application uses and is powered by Network File System (NFS).

* **BLOCK**: Block storage offers additional performance and reliability for databases or similar workloads, but is only able to be attached to 1 container at a time. This is powered by iSCSI.

* **S3 Object Storage**: Object based storage that is available via a web based API instead of through a mounted directory. A common implementation of this is the Amazon S3 API. This allows remote access storage over the internet that does not require directly attaching to a running system.

## Storage services

We have access to the following storage services for the OpenShift platform.

### OpenShift Persistent Volumes (NetApp)

The primary storage solution for the OpenShift platform is backed by a NetApp storage appliance that provides both file and block storage types to the cluster.

> All NetApp storage classes support resizing (bigger only). You can start with a small volume and edit your Persisent Volume Claim (PVC) to have a larger `.spec.resources.requests.storage` later if you need more. You may need to restart the pods attached to let the resize trigger on re-mount.

Reviewer: ```suggestion
All NetApp storage classes support resizing (bigger only). You can start with a small volume and edit your Persisent Volume Claim (PVC) to have a larger `.spec.resources.requests.storage` later if you need more. You may need to restart the pods attached to let the resize trigger on re-mount. **Note: the storage increase is capped by the storage quota assigned to the namespace. Check the current resource quota sizes [here](https://developer.gov.bc.ca/Project-Resource-Quotas-in-BC-Gov's-PaaS-(Openshift-4-Platform)).
```

* **NetApp Block**: `netapp-block-standard` is the current block `storageClass` target and the `storageClass` you should use for your block storage needs.

Two additional storage classes (`netapp-file-extended` and `netapp-block-extended`) aren't available to dev teams. They are reserved for operations teams to use for infrastructure components, like the ElasticSearch stack or the image registry.

### OpenShift Volume Snapshots

Snapshots can be used to maintain point-in-time copies of volumes. This empowers rapid testing by quickly resetting a PVC to a consistent state between each test run or getting an atomic view of a volume to run a backup on. Under the hood, they are very quickly created by the NetApp as a copy on write clone which means they are very space efficient.

See the [Red Hat docs](https://access.redhat.com/documentation/en-us/red_hat_openshift_container_storage/4.7/html/deploying_and_managing_openshift_container_storage_using_red_hat_openstack_platform/volume-snapshots_osp) for how to create snapshots and how to create PVCs from those snapshots.

### S3-Compatible Object Storage (Dell EMC Elastic Cloud Storage)

> For applications that need to store large amounts of data (beyond the storage allocation included in the medium size quota as defined [here](https://app.zenhub.com/workspaces/openshift-4-build-out-5db73142897668000144f22b/issues/bcdevops/openshift4-rollout/254)) we recommend the S3-compliant on-prem Object Storage Service offered by OCIO's Hosting Branch to store unstructured data such as images, pdfs, and other types of files. NetApp storage offered on the OCP4 Platform isn't suitable large amounts of unstructured data. NetApp storage should only be used for structured data such as databases that require high-I/O workloads.

Reviewer: ```suggestion
For applications that need to store large amounts of unstructured data we recommend using the S3-compliant on-prem [Object Store Service](https://ssbc-client.gov.bc.ca/services/ObjectStorage/overview.htm) offered by OCIO's Enterprise Hosting Branch. The Object Store Services provides storage for unstructured data such as images, pdfs, and other types of files. NetApp storage offered on the OCP4 Platform isn't suitable large amounts of unstructured data. NetApp storage should only be used for structured data such as databases that require high-I/O workloads.
```

## Tools

You can use tools to manage your persistent storage beyond the features built in to OpenShift.

* **Database backups**: We have a community project to help teams implement regular backups of their databases hosted within the platform. You can find the repo here: [BCDevOps/Backup-Container](https://github.com/bcdevops/backup-container)

* **Migrating storage**: another community supported repository is available to help with migrating data from one PVC to another (moving from one storageClass to another, moving to a larger PVC, etc). Please check out the repo here: [BCDevOps/StorageMigration](https://github.com/BCDevOps/StorageMigration)

## Storage details

### Quotas

> All storage sizes are in GiB and backup quotas default to half the storage size. These quotas can be requested in the Project Registry. You need to provide proof of increased storage needs to the Platform Product Director before the request can be approved and more storage is allocated to a namespace. All storage quotas include 60 PVCs and 5 VolumeSnapshots.

Reviewer: ```suggestion
All storage sizes are in GiB and backup quotas default to half the storage size. These quotas can be requested in the [Platform Project Registry](https://registry.developer.gov.bc.ca/public-landing) by a project's Product Owner or the Technical Lead. **Note: You need to provide proof of increased storage needs to the Platform Product Director before the request can be approved and more storage is allocated to a namespace.** All storage quotas include 60 PVCs and 5 VolumeSnapshots.
```

---
title: Platform storage

description: Describes the different storage technologies used on the OpenShift platform.

keywords: netapp, net app, nfs, storage, pvc, persistent, storage options

page purpose: Describes different platform technologies, types, and services and their details.

audience: technical lead, developer

author: Jonathan Bond

content owner: Olena Mitovska
---

# Platform storage

We use several different storage technologies on the OpenShift platform. Currently, we have three types of persistent storage available:

* **FILE**: File storage is an all-purpose storage type that can be attached to one or more containers. It's the recommended storage class for most general application uses and is powered by Network File System (NFS).

> * **BLOCK**: Block storage offers additional performance and reliability for databases or similar workloads, but is only able to be attached to 1 container at a time. This is powered by iSCSI.

Reviewer 2: Heads, up this isn't true! You can attached `netapp-block-standard` to more than one pod if you want to. The ability to mount a PV to multiple pods is set in your config, by specifying if the PV should be Read-Write-Once (RWO) or Read-Write-Many (RWX). The file/block storage classes have nothing to do with this!
